File: source/c/code/math/algebra/matrix/ntm_matrix_inverse.c
///////////////////////////////////////////////////////////////////////////////////
//                                            __ _      _     _                  //
//                                          /_(_)    | |   | |                 //
//                __ _ _   _  ___  ___ _ __ | |_ _  ___| | __| |                 //
//              /_` | | | |/ _ \/ _ \ '_ \|  _| |/ _ \ |/ _` |                 //
//              | (_| | |_| |  __/  __/ | | | | | |  __/ | (_| |                 //
//               \__, |\__,_|\___|\___|_| |_|_| |_|\___|_|\__,_|                 //
//                  | |                                                          //
//                  |_|                                                          //
//                                                                               //
//                                                                               //
//              Peripheral-NTM for MPSoC                                         //
//              Neural Turing Machine for MPSoC                                  //
//                                                                               //
///////////////////////////////////////////////////////////////////////////////////

///////////////////////////////////////////////////////////////////////////////////
//                                                                               //
// Copyright (c) 2020-2024 by the author(s)                                      //
//                                                                               //
// Permission is hereby granted, free of charge, to any person obtaining matrix copy  //
// of this software and associated documentation files (the "Software"), to deal //
// in the Software without restriction, including without limitation the rights  //
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell     //
// copies of the Software, and to permit persons to whom the Software is         //
// furnished to do so, subject to the following conditions:                      //
//                                                                               //
// The above copyright notice and this permission notice shall be included in    //
// all copies or substantial portions of the Software.                           //
//                                                                               //
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR    //
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,      //
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE   //
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER        //
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, //
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN     //
// THE SOFTWARE.                                                                 //
//                                                                               //
// ============================================================================= //
// Author(s):                                                                    //
//   Paco Reina Campo <pacoreinacampo@queenfield.tech>                           //
//                                                                               //
///////////////////////////////////////////////////////////////////////////////////

#include <stdio.h>
#include <stdlib.h>
#include <assert.h>

#define SIZE 10

#define SIZE_IN 3

double ntm_matrix_inverse(double **data_in) {
  double matrix[SIZE][SIZE];

  double **data_out;

  double ratio;

  int i, j, m;

  data_out = (double **) malloc(SIZE_IN*sizeof(int*));

  for (i = 0; i < SIZE_IN; i++) {
    data_out[i] = (double *)malloc(SIZE_IN*sizeof(int));
  }

  // Augmenting Identity Matrix of Order SIZE_IN
  for (i = 0; i < SIZE_IN; i++) {
    for (j = 0; j < SIZE_IN; j++) {
      matrix[i][j] = data_in[i][j];

      if (i == j) {
        matrix[i][j + SIZE_IN] = 1;
      } else {
        matrix[i][j + SIZE_IN] = 0;
      }
    }
  }

  // Applying Gauss Jordan Elimination
  for (i = 0; i < SIZE_IN; i++) {
    if (matrix[i][i] == 0.0) {
      printf("Mathematical Error!");
    }
    for (j = 0; j < SIZE_IN; j++) {
      if (i != j) {
        ratio = matrix[j][i]/matrix[i][i];
        for (m = 1; m <= 2*SIZE_IN; m++) {
          matrix[j][m] = matrix[j][m] - ratio*matrix[i][m];
        }
      }
    }
  }

  // Row Operation to Make Principal Diagonal to 1
  for (i = 0; i < SIZE_IN; i++) {
    for (j = SIZE_IN; j < 2*SIZE_IN; j++) {
      matrix[i][j] = matrix[i][j]/matrix[i][i];
    }
  }

  // Output
  for (i = 0; i < SIZE_IN; i++) {
    for (j = 0; j < SIZE_IN; j++) {
      data_out[i][j] = matrix[i][j + SIZE_IN];
    }
  }

  return **data_out;
}

int main() {
  double **data_in;

  double **data_out;

  int i;

  data_in = (double **) malloc(SIZE_IN*sizeof(int*));

  data_out = (double **) malloc(SIZE_IN*sizeof(int*));

  for (i = 0; i < SIZE_IN; i++) {
    data_in[i] = (double *)malloc(SIZE_IN*sizeof(int));

    data_out[i] = (double *)malloc(SIZE_IN*sizeof(int));
  }

  data_in[0][0] = 1.0; data_in[1][0] =  1.0; data_in[2][0] = -2.0;
  data_in[0][1] = 1.0; data_in[1][1] =  3.0; data_in[2][1] = -4.0;
  data_in[0][2] = 3.0; data_in[1][2] = -3.0; data_in[2][2] = -4.0;

  data_out[0][0] =  3.00; data_out[1][0] =  1.00; data_out[2][0] =  1.50;
  data_out[0][1] = -1.25; data_out[1][1] = -0.25; data_out[2][1] = -0.75;
  data_out[0][2] = -0.25; data_out[1][2] = -0.25; data_out[2][2] = -0.25;

  assert(ntm_matrix_inverse(data_in)==**data_out);

  free(data_in);

  free(data_out);

  return 0;
}

File: source/rust/code/math/algebra/matrix/ntm_matrix_inverse.rs
///////////////////////////////////////////////////////////////////////////////////
//                                            __ _      _     _                  //
//                                           / _(_)    | |   | |                 //
//                __ _ _   _  ___  ___ _ __ | |_ _  ___| | __| |                 //
//               / _` | | | |/ _ \/ _ \ '_ \|  _| |/ _ \ |/ _` |                 //
//              | (_| | |_| |  __/  __/ | | | | | |  __/ | (_| |                 //
//               \__, |\__,_|\___|\___|_| |_|_| |_|\___|_|\__,_|                 //
//                  | |                                                          //
//                  |_|                                                          //
//                                                                               //
//                                                                               //
//              Peripheral-NTM for MPSoC                                         //
//              Neural Turing Machine for MPSoC                                  //
//                                                                               //
///////////////////////////////////////////////////////////////////////////////////

///////////////////////////////////////////////////////////////////////////////////
//                                                                               //
// Copyright (c) 2020-2024 by the author(s)                                      //
//                                                                               //
// Permission is hereby granted, free of charge, to any person obtaining a copy  //
// of this software and associated documentation files (the "Software"), to deal //
// in the Software without restriction, including without limitation the rights  //
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell     //
// copies of the Software, and to permit persons to whom the Software is         //
// furnished to do so, subject to the following conditions:                      //
//                                                                               //
// The above copyright notice and this permission notice shall be included in    //
// all copies or substantial portions of the Software.                           //
//                                                                               //
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR    //
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,      //
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE   //
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER        //
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, //
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN     //
// THE SOFTWARE.                                                                 //
//                                                                               //
// ============================================================================= //
// Author(s):                                                                    //
//   Paco Reina Campo <pacoreinacampo@queenfield.tech>                           //
//                                                                               //
///////////////////////////////////////////////////////////////////////////////////

// Matrix Inversion
pub fn ntm_matrix_inverse(matrix: &mut Vec<Vec<f64>>) -> Vec<Vec<f64>>{
    let length = matrix.len();

    let mut augmented_matrix = ntm_zero_matrix(length, 2*length);

    for i in 0..length {
        for j in 0.. length {
            augmented_matrix[i][j] = matrix[i][j];
        }
        augmented_matrix[i][i + length] = 1.0;
    }

    ntm_gauss_jordan_general(&mut augmented_matrix);

    let mut data_out = ntm_zero_matrix(length, length);

    for i in 0..length {
        for j in 0..length {
            data_out[i][j] = augmented_matrix[i][j+length];
        }
    }
    data_out
}

// Generalised Reduced Row Echelon Form
pub fn ntm_gauss_jordan_general(matrix: &mut Vec<Vec<f64>>) {
    let mut lead = 0;

    let i_counter = matrix.len();
    let j_counter = matrix[0].len();

    for r in 0..i_counter {
        if j_counter <= lead {
            break;
        }

        let mut i = r;

        while matrix[i][lead] == 0.0 {
            i = i + 1;

            if i_counter == i {
                i = r;
                lead = lead + 1;

                if j_counter == lead {
                    break;
                }
            }
        }

        let temporal = matrix[i].to_owned();

        matrix[i] = matrix[r].to_owned();
        matrix[r] = temporal.to_owned();

        if matrix[r][lead] != 0.0 {
            let divider_ratio = matrix[r][lead];

            for j in 0..j_counter {
                matrix[r][j] = matrix[r][j] / divider_ratio;
            }
        }

        for k in 0..i_counter {
            if k != r {
                let multiplier_ratio = matrix[k][lead];

                for j in 0..j_counter {
                    matrix[k][j] = matrix[k][j] - matrix[r][j] * multiplier_ratio;
                }
            }
        }
        lead = lead + 1;
    }
}

// Zero Matrix
pub fn ntm_zero_matrix(size_i_in: usize, size_j_in: usize) -> Vec<Vec<f64>> {
    let mut matrix = Vec::with_capacity(size_i_in);

    for _ in 0..size_i_in {
        let mut vector: Vec<f64> = Vec::with_capacity(size_j_in);

        for _ in 0..size_j_in {
            vector.push(0.0);
        }
        matrix.push(vector);
    }
    matrix
}

// Printed Matrix
pub fn ntm_print_matrix(matrix: &mut Vec<Vec<f64>>) {
    for i in 0..matrix.len(){
        println!("{:?}", matrix[i]);
    }
}

fn main() {
    let mut data_in_0: Vec<Vec<f64>> = vec![
        vec![1.0, 2.0, 3.0],
        vec![4.0, 1.0, 6.0],
        vec![7.0, 8.0, 0.0]
    ];
    let mut data_in_1: Vec<Vec<f64>> = vec![
        vec![ 0.0, -1.0,  0.0],
        vec![-1.0,  2.0, -1.0],
        vec![ 0.0, -1.0,  2.0]
    ];

    let data_out_0: Vec<Vec<f64>> = vec![
        vec![-0.43243243243243240,  0.21621621621621620,  0.08108108108108109],
        vec![ 0.37837837837837834, -0.18918918918918917,  0.05405405405405404],
        vec![ 0.22522522522522523,  0.05405405405405405, -0.06306306306306306]
    ];
    let data_out_1: Vec<Vec<f64>> = vec![
        vec![-1.5, -1.0, -0.5],
        vec![-1.0,  0.0,  0.0],
        vec![-0.5,  0.0,  0.5]
    ];

    let mut reference_0 = &mut data_in_0;
    let mut reference_1 = &mut data_in_1;

    let double_reference_0 = &mut reference_0;
    let double_reference_1 = &mut reference_1;

    println!("Matrix 0:\n");
    ntm_print_matrix(double_reference_0);
    println!("\nInverse of Matrix 0:\n");
    ntm_print_matrix(&mut ntm_matrix_inverse(double_reference_0));

    println!("\n\nMatrix 1:\n");
    ntm_print_matrix(double_reference_1);
    println!("\nInverse of Matrix 1:\n");
    ntm_print_matrix(&mut ntm_matrix_inverse(double_reference_1));

    assert_eq!(ntm_matrix_inverse(double_reference_0), data_out_0);
    assert_eq!(ntm_matrix_inverse(double_reference_1), data_out_1);
}
